### PR TITLE
[Android] Remove deprecated APIs and fix for a few found issues

### LINKED ISSUE #32: FlutterSegmentPlugin.java uses or overrides a deprecated API.
We have a deprecated implementation for the plugin on android:

There's a warning running the project:
```dart
android/src/main/java/com/example/flutter_segment/FlutterSegmentPlugin.java uses or overrides a deprecated API.
```

### PR DESCRIPTION
**Pull request by vbuberen**
This PR mainly addresses and closes #32

Changes:

- Replaced deprecated `middleware()` from Segment Android with suggested `useSourceMiddleware()`
- Removed `registerWith()` as it is also deprecated and was needed only when Flutter Embedding V1 was used, while the plugin already switched to V2, so register will happen automatically.
- A few fixes for lint warnings, like adding correct annotations, switch to lambda in `userSourceMiddleware()`, etc.
- Add a teardown for `methodChannel` in `onDetachFromEngine` as it should be done.
- Bumped Android Gradle plugin to `7.2.1` as [it was a hotfix](https://developer.android.com/studio/releases/gradle-plugin) for `7.2.0` 
- Updated example to not use `jcenter`.

### CODE REVIEW

File: example/android/build.gradle [hunk 1]
```diff
@@ -1,18 +1,18 @@
...
-        classpath 'com.android.tools.build:gradle:7.2.0'
+        classpath 'com.android.tools.build:gradle:7.2.1'
     }
 }
 
 allprojects {
     repositories {
         google()
-        jcenter()
+        mavenCentral()
```
Comment (danielgomezrico) [inline example/android/build.gradle:15]: Why do we need to update this?

**Review by danielgomezrico — APPROVED** [review]
LGTM, thanks for this! let me do some tests before the merge

File: example/android/build.gradle (hunk 1 above)
Comment (vbuberen) [reply example/android/build.gradle:15]: You mean why we need to replace `jcenter` with `mavenCentral`?

The replacement is needed to have more stable place to get project dependencies (for example, Kotlin, which is required by Android Gradle plugin, etc.). Also mentioned the issue with `jcenter` in https://github.com/la-haus/flutter-segment/pull/27#pullrequestreview-971439115

Last time `jcenter` was down there were no communication from Bintray on when it will be fixed and it took quite some time, so I think it is not good to continue relying on it even for example apps.

Comment (danielgomezrico) [reply example/android/build.gradle:15]: Got it, thanks for the answer

**Comment by danielgomezrico** [comment]
@vbuberen I tried and on compilation it fails with:

```kotlin
/Users/daniel/.pub-cache/git/flutter-segment-95ad9e4b3158442f3dd78f6d0dade7b03b8a2875/android/src/main/java/com/example/flutter_segment/FlutterSegmentPlugin.java:90: error: lambda expressions are not supported in -source 7
              chain -> {
                    ^
  (use -source 8 or higher to enable lambda expressions)
```

File: android/src/main/java/com/example/flutter_segment/FlutterSegmentPlugin.java [hunk 2]
```diff
@@ -94,34 +86,31 @@ private void setupChannels(FlutterSegmentOptions options) {
...
-      analyticsBuilder.middleware(
-        new Middleware() {
-          @Override
-          public void intercept(Chain chain) {
-            try {
-              if (appendToContextMiddleware == null) {
-                chain.proceed(chain.payload());
-                return;
+      analyticsBuilder.useSourceMiddleware(
+              chain -> {
```
Comment (cdmunoz) [inline android/src/main/java/com/example/flutter_segment/FlutterSegmentPlugin.java]: Is this supported by the current JDK? I believe this forces to use a Java 8. I believe this Java version should not be supported yet, and it could be part of another PR. The first thing here is to make this Java 7 compatible.

**Review by cdmunoz — CHANGES_REQUESTED** [review]
LGTM 👌🏻

Although it should not be upgraded to Java 8

File: android/src/main/java/com/example/flutter_segment/FlutterSegmentPlugin.java (hunk 2 above)
Comment (vbuberen) [reply android/src/main/java/com/example/flutter_segment/FlutterSegmentPlugin.java]: I will make usage of Java 8 in a different PR, but I don't think that project should be strictly relying only on Java 7 language features, because de-sugaring was implemented quite a long time ago already in Android Gradle plugin, so there are no issues with using Java 8 language features in Android projects supporting very old APIs, like this plugin does with using minSDK 16.

Here is, for example, an article from a few years ago https://developer.android.com/studio/write/java8-support

**Comment by vbuberen** [comment]
> @vbuberen I tried and on compilation it fails with:
> 
> ```kotlin
> /Users/daniel/.pub-cache/git/flutter-segment-95ad9e4b3158442f3dd78f6d0dade7b03b8a2875/android/src/main/java/com/example/flutter_segment/FlutterSegmentPlugin.java:90: error: lambda expressions are not supported in -source 7
>               chain -> {
>                     ^
>   (use -source 8 or higher to enable lambda expressions)
> ```

Looks like I needed to explicitly specify in build.gradle the following block:

```
compileOptions {
        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8
    }
```

In latest Android Studio versions it is not an issue as project builds fine even without that part I mentioned above.

File: android/src/main/java/com/example/flutter_segment/FlutterSegmentPlugin.java (hunk 2 above)
Comment (vbuberen) [reply android/src/main/java/com/example/flutter_segment/FlutterSegmentPlugin.java]: Reverted usage of Java 8 language features.

**Review by cdmunoz — APPROVED** [review]
Awesome job! 
Thank you very much for your contribution!

**Review by juansoos — APPROVED** [review]
